trace-utils/src/msgpack_decoder/v04/number.rs

```diff
@@ -0,0 +1,236 @@
+// Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use super::error::DecodeError;
+use rmp::{decode::RmpRead, Marker};
+use std::fmt;
+
+pub enum Number {
+    U8(u8),
+    U32(u32),
+    U64(u64),
+    I8(i8),
+    I32(i32),
+    I64(i64),
+    F64(f64),
+}
```

Comment:
I don't really see why you need this runtime representation of how the message pack is encoded?

You could make the code simpler by having two read function
* read_number_signed<T: TryFrom<i64>>() ->  Result<T> 
* read_number_unsigned<T: TryFrom<u64>>() ->  Result<T> 

What would these functions do is:
* Widening the decoded numbers to i64 or u64, since these type would be big enough to represent any signed or unsigned number using the existing [TryFrom<number>  trait implementation](https://doc.rust-lang.org/std/convert/trait.TryFrom.html#implementors) for these, then call T::try_from to cast to the desired output.

See this playground for a small implementation of this idea https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2dfcc2b984bcebb4395179a62087c97f

Reply:
If I understand your suggestion correctly we'd change `Number` to be:
```rust
pub enum Number {
    Signed(i64),
    Unsigned(u64),
    Float(f64),
}
```

The existing `read_number` function wouldn't be simplified much because we still need to handle all the different `Marker` types. It would wind up looking like:

```rust
pub fn read_number(buf: &mut &[u8]) -> Result<Number, DecodeError> {
    match rmp::decode::read_marker(buf)
        .map_err(|_| DecodeError::InvalidFormat("Unable to read marker for number".to_owned()))?
    {
        Marker::FixPos(val) => Ok(Number::Unsigned(val as u64)),
        Marker::FixNeg(val) => Ok(Number::Signed(val as i64)),
        Marker::U8 => Ok(Number::Unsigned(
            buf.read_data_u8().map_err(|_| DecodeError::IOError)? as u64,
        )),
        Marker::U16 => Ok(Number::Unsigned(
            buf.read_data_u16().map_err(|_| DecodeError::IOError)? as u64,
        )),
        Marker::U32 => Ok(Number::Unsigned(
            buf.read_data_u32().map_err(|_| DecodeError::IOError)? as u64,
        )),
        Marker::U64 => Ok(Number::Unsigned(
            buf.read_data_u64().map_err(|_| DecodeError::IOError)?,
        )),
        Marker::I8 => Ok(Number::Signed(
            buf.read_data_i8().map_err(|_| DecodeError::IOError)? as i64,
        )),
        Marker::I16 => Ok(Number::Signed(
            buf.read_data_i16().map_err(|_| DecodeError::IOError)? as i64,
        )),
        Marker::I32 => Ok(Number::Signed(
            buf.read_data_i32().map_err(|_| DecodeError::IOError)? as i64,
        )),
        Marker::I64 => Ok(Number::Signed(
            buf.read_data_i64().map_err(|_| DecodeError::IOError)?,
        )),
        Marker::F32 => Ok(Number::Float(
            buf.read_data_f32().map_err(|_| DecodeError::IOError)? as f64,
        )),
        Marker::F64 => Ok(Number::Float(
            buf.read_data_f64().map_err(|_| DecodeError::IOError)?,
        )),
        _ => Err(DecodeError::InvalidType("Invalid number type".to_owned())),
    }
}
```

A bigger problem arises when we implement the `TryFrom` traits and need to downcast.

The following will silently truncate a u64 value to u32:

```rust
impl TryFrom<Number> for u32 {
    type Error = DecodeError;
    fn try_from(value: Number) -> Result<Self, Self::Error> {
        match value {
            Number::Unsigned(val) => Ok(val as u32),
            _ => Err(DecodeError::InvalidConversion(format!(
                "unable to convert {} to u32",
                value
            ))),
        }
    }
}
```

To raise an error when we can't downcast we wind up with:

```rust
impl TryFrom<Number> for u32 {
    type Error = DecodeError;
    fn try_from(value: Number) -> Result<Self, Self::Error> {
        match value {
            Number::Unsigned(val) if val <= u32::MAX as u64 => Ok(val as u32),
            _ => Err(DecodeError::InvalidConversion(format!(
                "unable to convert {} to u32",
                value
            ))),
        }
    }
}
```

I lean slightly towards preferring the existing implementation as it seems a bit clearer to me. @hoolioh you originally implemented this. Do you have any thoughts?

And, apologies if I misunderstood your suggestion.